networkit/cpp/community/ParallelLeiden.cpp

```diff
@@ -0,0 +1,530 @@
+                result = std::move(p);
+                std::swap(temp, coarse);
+                currentGraph = &coarse;
+            }
+        } while (!done);
+        flattenPartition();
+        INFO("Leiden iteration done, took " + totalTime.elapsedTag() + "so far");
+    } while (changed && numberOfIterations > 0);
+    hasRun = true;
+}
+
+void ParallelLeiden::calculateVolumes(const Graph &graph) {
+    auto timer = Aux::Timer();
+    timer.start();
+    // thread safe reduction. Avoid atomic calculation of total graph volume for unweighted graphs.
+    // Vol(G) is then 2*|E|
+    communityVolumes.clear();
+    communityVolumes.resize(result.upperBound() + 10000);
+    if (graph.isWeighted()) {
+        std::vector<double> threadVolumes(omp_get_max_threads());
+        graph.parallelForNodes([&](node a) {
+            {
+                edgeweight ew = graph.weightedDegree(a, true);
+#pragma omp atomic
+                communityVolumes[result[a]] += ew;
+                threadVolumes[omp_get_thread_num()] += ew;
+            }
+        });
+        for (const auto vol : threadVolumes) {
+            inverseGraphVolume += vol;
+        }
+        inverseGraphVolume = 1 / inverseGraphVolume;
+    } else {
+        inverseGraphVolume = 1.0 / (2 * graph.numberOfEdges());
+        graph.parallelForNodes([&](node a) {
+            {
+#pragma omp atomic
+                communityVolumes[result[a]] += graph.weightedDegree(a, true);
+            }
+        });
+    }
+    TRACE("Calculating Volumes took " + timer.elapsedTag());
+}
+
+void ParallelLeiden::flattenPartition() {
+    auto timer = Aux::Timer();
+    timer.start();
+    if (mappings.empty()) {
+        return;
+    }
+    // Create a new partition with size |V(G)| (the fine/bigger Graph)
+    Partition flattenedPartition(G->numberOfNodes());
+    flattenedPartition.setUpperBound(result.upperBound());
+    int i = mappings.size() - 1;
+    std::vector<node> &lower = mappings[i--];
+    while (i >= 0) {
+        // iteratively "resolve" (i.e compose) mappings. Let "lower" be a mapping thats
+        // below "higher" in the hierarchy (i.e. of a later aggregation)
+        // If higher[index] = z and lower[z] = x then set higher[index] = x
+        std::vector<node> &upper = mappings[i--];
+        for (auto &idx : upper) {
+            idx = lower[idx];
+        }
+        lower = upper;
+    }
+    G->parallelForNodes([&](node a) { flattenedPartition[a] = lower[a]; });
+    flattenedPartition.compact(true);
+    result = flattenedPartition;
+    mappings.clear();
+    TRACE("Flattening partition took " + timer.elapsedTag());
+}
+
+void ParallelLeiden::parallelMove(const Graph &graph) {
+    DEBUG("Local Moving : " + std::to_string(graph.numberOfNodes()) + " Nodes ");
+    std::vector<count> moved(omp_get_max_threads(), 0);
+    std::vector<count> totalNodesPerThread(omp_get_max_threads(), 0);
+    std::atomic_int singleton(0);
+    //^^^ Debug stuff
```

Comment:
Are these only needed for debug purposes right? Why don't you use them only in debug mode with `#ifndef NDEBUG`?

Reply:
Mainly because of readability. I used to have these enclosed by ifndef's but I'd also have to enclose all the increments in the middle of the code.

If you don't care about that, I'll enclose them again.

Reply:
Namely, these two:
```if (0 > modThreshold || maxDelta > modThreshold) {
                    moved[omp_get_thread_num()]++;
                    if (0 > maxDelta) { // move node to empty community
                        singleton++;
```
Thats why i used to have a macro that only inserted that code when ndebug wasn't present (which made it look like a function call / a one liner, instead of the one line + two additional ones)

Reply:
I see the readability argument. It might be a good idea to add a similar macro to NetworKit instead of defining it for just one class.

Reply:
Perhaps, This would probably belong to a new pull request though. 
Where do you suggest it to be put then, and it's name?
Simply DEBUGONLY ?

Reply:
Yes, this should be implemented in another PR. The details about how to implement this can be discussed in a new issue/discussion.

Reply:
As discussed in #839, I take back my comment. It should not be too costly anyway to have those vectors in release mode as well.